### PR TITLE
Refactor sharable logic in gatsby-node

### PR DESCRIPTION
**Pull request by jasonbcox**
This is in preparation for new `createPages` logic that will generate some pages and/or redirects based on frontmatter. `createPages` can share some of the localization logic used by `onCreatePage`, so this patch refactors to make it available.

There is no change in behavior.

### CODE REVIEW

File: v2/gatsby-node.js [hunk 1]
```diff
@@ -1,31 +1,44 @@
...
+async function createPageForAllLocales(createPage, page, createFn) {
   await Promise.all(
     Object.entries(locales).map(async ([locale, params]) => {
       const originalPath = page.path
+      const pathWithLocale = localizePath(originalPath, params)
 
-      let localizedPath = originalPath
-      if (params.slug) {
-        localizedPath = "/" + params.slug + localizedPath
+      if (createFn) {
```
Comment (deadalnix) [inline v2/gatsby-node.js:18]: No need to branch here.

File: v2/gatsby-node.js [hunk 2]
```diff
@@ -1,31 +1,44 @@
...
+  return path
+}
 
-  // Create one page for each locale
+// createFn is an optional override of the default createPage behavior
+async function createPageForAllLocales(createPage, page, createFn) {
   await Promise.all(
     Object.entries(locales).map(async ([locale, params]) => {
       const originalPath = page.path
+      const pathWithLocale = localizePath(originalPath, params)
```
Comment (deadalnix) [inline v2/gatsby-node.js:16]: That would be the `loalizedPath` ?

File: v2/gatsby-node.js [hunk 3]
```diff
@@ -1,31 +1,44 @@
...
-}) => {
-  // Delete the original page (since we are gonna create localized versions of it)
-  await deletePage(page)
+// Prepend the localization slug, if there is one (English does not have one)
+function localizePath(path, params) {
+  if (params.slug) {
+    path = "/" + params.slug + path
+  }
+  return path
+}
```
Comment (deadalnix) [inline v2/gatsby-node.js:9]: It seems doubtful that this is useful on its own. This most likely translate a design problem in what you got coming down the road.

File: v2/gatsby-node.js [hunk 4]
```diff
@@ -1,31 +1,44 @@
...
 
-  // Create one page for each locale
+// createFn is an optional override of the default createPage behavior
+async function createPageForAllLocales(createPage, page, createFn) {
```
Comment (deadalnix) [inline v2/gatsby-node.js:12]: Use the lambda style syntax like in the rest of the codebase.

**Review by deadalnix — CHANGES_REQUESTED** [review]
This is a good idea to start to extract the components that work on their own.

I think think you current decoupage illustrate pretty well why your previous approach was poorly structured. Especially the branching in the middle of `createPageForAllLocales` . I'm fairly confident that you only felt the need to extract `localizePath` as a workaround for this choice.

File: v2/gatsby-node.js (hunk 3 above)
Comment (jasonbcox) [reply v2/gatsby-node.js:9]: I need a way to localize the v1 path override. An alternative would be to pass a localization prefix to createFn, but this makes a stronger assumption that localized paths will always be prefix-only.

Comment (deadalnix) [reply v2/gatsby-node.js:9]: You don't if you start from the localized path remove the trailing slash and add ".html"

File: v2/gatsby-node.js (hunk 1 above)
Comment (deadalnix) [reply v2/gatsby-node.js:18]: This still does not make sense.

**Review by deadalnix — CHANGES_REQUESTED** [review]

File: v2/gatsby-node.js (hunk 3 above)
Comment (jasonbcox) [reply v2/gatsby-node.js:9]: ...localize the v1 path **override**, not the default behavior.

File: v2/gatsby-node.js (hunk 2 above)
Comment (jasonbcox) [reply v2/gatsby-node.js:16]: It looks too similar to the function name `localizePath` so I renamed it to another appropriate name.
Alternative is `const localizedPath = addLocaleToPath(...)`

Comment (jasonbcox) [reply v2/gatsby-node.js:16]: This naming is actually strictly more correct, so let's go with that.

File: v2/gatsby-node.js (hunk 3 above)
Comment (deadalnix) [reply v2/gatsby-node.js:9]: Do we actually need a v1 path override? Can we add it if/when we need it?

Comment (deadalnix) [reply v2/gatsby-node.js:9]: This ^

**Review by deadalnix — CHANGES_REQUESTED** [review]

File: v2/gatsby-node.js (hunk 3 above)
Comment (jasonbcox) [reply v2/gatsby-node.js:9]: We can add it when needed. The whole point of adding it up-front was to ensure low-friction for fixing routing issues that may appear in production.

Comment (deadalnix) [reply v2/gatsby-node.js]: That's also how you end up with this marvel: https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/aop/framework/AbstractSingletonProxyFactoryBean.html

You'll note that it makes things easier if things evolve in a way you expected, but also makes things worse when thing evolve in a way you did not expect.

**Pull request merged**